flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/BoundedProcessingOverWindowFunction.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.runtime.aggregate
+
+import org.apache.flink.api.java.tuple.Tuple
+import org.apache.flink.types.Row
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.streaming.api.functions.windowing.RichWindowFunction
+import org.apache.flink.streaming.api.windowing.windows.Window
+import org.apache.flink.util.Collector
+import org.apache.flink.table.functions.AggregateFunction
+import org.apache.flink.api.java.typeutils.RowTypeInfo
+import org.apache.flink.util.Preconditions
+import org.apache.flink.table.functions.Accumulator
+import java.lang.Iterable
+
+class BoundedProcessingOverWindowFunction[W <: Window](
```

Comment:
As @huawei-flink pointed out, I agreed to use a WindowFunction for the bounded OVER ROW case in our previous discussions. Although it would not be consistent with the other OVER windows, this would be an easier solution, IMO.

However, I think we have to use a `ProcessFunction` for correctness reasons. An OVER aggregation must emit one row for each input row. If we have an `OVER (PARTITION BY a ORDER BY procTime() ROWS BETWEEN 3 PRECEDING AND CURRENT ROW)` and implement this with a `SlidingCountWindow(3, 1)` we will lose the two rows because the window function is called the first time, when three elements have arrived.

Regarding the aggregation of COUNT DISTINCT, we could implement a custom aggregation function, similar to the retractable min or max aggregation function.